packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -174,8 +174,33 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>)
       lastValue.current = inputValue;
     }
 
+    // Update the inputValue if the selected item's text differs from the current input value but only if items are controlled.
+    // This is to handle cases where a selectedKey is specified but the items aren't available (async loading), the selected item's text value updates.
+    // Only reset if the user isn't currently within the field so we don't erroneously modify user input.
+    // Trigger only if a selectedKey exists so that allowCustomValue=true user values don't get reset.
+    if (!isFocused && props.items !== lastItems.current && selectedKey != null) {
+      let itemText = collection.getItem(selectedKey)?.textValue ?? '';
+      if (inputValue !== itemText) {
+        if (
+          (props.selectedKey !== undefined && props.inputValue !== undefined) ||
+          (props.inputValue !== undefined && props.isOpen !== undefined)
```

Comment:
Similar to the `commitSelection` logic, but w/o the `selectedKey + isOpen` controlled condition here since `resetInputValue` should work just fine on its own and we don't need to explicitly call `onSelectionChange` to close the menu IMO. Still mulling this over

Reply:
Hmm, should we handle this case at all or should we only handle when `inputValue` is uncontrolled? I feel like maybe it's the user's responsibility to update `inputValue` when they update `items` if it is controlled.

Reply:
What about when someone uses `useAsyncList` with the ComboBox? Would this be the expectation:
-  they calculate the new `filterValue` from the fetched list of items inside the `load` func that they provide to the `useAsyncList` hook 
- they call `list.setFilterText` or provide the adjusted `filterText` value through `load`'s return object?

Just trying to feel out the mechanics of how the user would go about doing this. IMO I felt this was quite similar to our `commitSelection` logic in that we reset the `inputValue` (or call onSelectionChange instead, depending on what is controlled) for them even if `inputValue` was controlled

Reply:
Well, feels like if you're controlling `inputValue` you wouldn't necessarily want it to change, even when `items` changes, and therefore it's your responsibility to update it in that case. Basically, whenever you control `inputValue` you need to update it whenever anything changes that would affect it, and I think that's true in other cases (e.g. in `onSelectionChange`). So yeah, they'd need to do it in the places you said.

Reply:
So essentially the logic would become just:
```
 if (props.inputValue === undefined) {
     resetInputValue();
 }
```
?

I guess this is fine as long as we call this out in the docs (aka if users are controlling items or providing a item map then they should be aware that inputValue and selectedKey text value could drift).

Reply:
You're mentioning updating the docs, but I don't see that change? Separate PR based on how docs deploy?